Add retries when downloading nuget.exe

KoreBuild should add retries when downloading nuget.exe to prevent build breaks.

https://github.com/aspnet/KoreBuild/blob/dev/build/KoreBuild.ps1#L94-L99
https://github.com/aspnet/KoreBuild/blob/dev/build/KoreBuild.sh#L147-L151

Another failure: http://aspnetci/viewLog.html?buildId=206695&tab=buildResultsDiv&buildTypeId=XPlat_Windows_Win10Universe

```
  Invoke-WebRequest : Unable to connect to the remote server
  At C:\BuildAgent\work\2b0c8ce712895332\.r\KestrelHttpServer\.build\KoreBuild.ps1:101 char:5
  +     Invoke-WebRequest "https://dist.nuget.org/win-x86-commandline/v4. ...
  +     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebExc 
     eption
      + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
   
C:\BuildAgent\work\2b0c8ce712895332\build\RepositoryBuild.targets(60,5): error MSB3073: The command "./build.cmd /t:Verify /p:BuildNumber=24579 /p:Configuration=Release" exited with code 1.
```

--- a/.gitignore
+++ b/.gitignore
@@ -14,7 +14,8 @@ PublishProfiles/
 *.cache
 *.docstates
 _ReSharper.*
-nuget.exe
+*.exe
+*.dll
 *net45.csproj
 *net451.csproj
 *k10.csproj

--- a/build/KoreBuild.ps1
+++ b/build/KoreBuild.ps1
@@ -122,5 +122,5 @@ if (!(Test-Path $msbuildArtifactsDir))
 
 $msBuildArguments | Out-File -Encoding ASCII -FilePath $msBuildResponseFile
 
-exec dotnet msbuild /nologo /t:Restore /p:PreflightRestore=true "$makeFileProj"
+exec dotnet msbuild /nologo /t:Restore /p:PreflightRestore=true /p:RepositoryRoot="$repoFolder/" "$makeFileProj"
 exec dotnet msbuild `@"$msBuildResponseFile"

--- a/build/KoreBuild.ps1
+++ b/build/KoreBuild.ps1
@@ -98,30 +98,10 @@ if (!($env:Path.Split(';') -icontains $dotnetLocalInstallFolder))
     $env:Path = "$newPath"
 }
 
-# wokaround for CLI issue: https://github.com/dotnet/cli/issues/2143
-$sharedPath = (Join-Path (Split-Path ((get-command dotnet.exe).Path) -Parent) "shared");
-(Get-ChildItem $sharedPath -Recurse *dotnet.exe) | %{ $_.FullName } | Remove-Item;
-
-# We still nuget because dotnet doesn't have support for pushing packages
-$nugetExePath = Join-Path $PSScriptRoot 'nuget.exe'
-if (!(Test-Path $nugetExePath))
-{
-    Invoke-WebRequest "https://dist.nuget.org/win-x86-commandline/v4.0.0-rc4/NuGet.exe" -OutFile "$PSScriptRoot/nuget.exe"
-}
-
 $makeFileProj = "$PSScriptRoot/KoreBuild.proj"
 $msbuildArtifactsDir = "$repoFolder/artifacts/msbuild"
 $msbuildLogFilePath = "$msbuildArtifactsDir/msbuild.log"
 $msBuildResponseFile = "$msbuildArtifactsDir/msbuild.rsp"
-
-$preflightClpOption='/clp:DisableConsoleColor'
-$msbuildClpOption='/clp:DisableConsoleColor;Summary'
-if ("${env:CI}${env:APPVEYOR}${env:TEAMCITY_VERSION}${env:TRAVIS}" -eq "")
-{
-    # Not on any of the CI machines. Fine to use colors.
-    $preflightClpOption=''
-    $msbuildClpOption='/clp:Summary'
-}
 
 $msBuildArguments = @"
 /nologo
@@ -129,7 +109,7 @@ $msBuildArguments = @"
 /p:RepositoryRoot="$repoFolder/"
 /fl
 /flp:LogFile="$msbuildLogFilePath";Verbosity=detailed;Encoding=UTF-8
-$msbuildClpOption
+/clp:Summary
 "$makeFileProj"
 "@
 
@@ -142,7 +122,5 @@ if (!(Test-Path $msbuildArtifactsDir))
 
 $msBuildArguments | Out-File -Encoding ASCII -FilePath $msBuildResponseFile
 
-# workaround https://github.com/dotnet/core-setup/issues/1664
-"{ `"sdk`": { `"version`": `"$dotnetVersion`" } }" | Out-File "$repoFolder/global.json" -Encoding ascii
-exec dotnet msbuild /nologo $preflightClpOption /t:Restore /p:PreflightRestore=true "$makeFileProj"
+exec dotnet msbuild /nologo /t:Restore /p:PreflightRestore=true "$makeFileProj"
 exec dotnet msbuild `@"$msBuildResponseFile"

--- a/build/KoreBuild.ps1
+++ b/build/KoreBuild.ps1
@@ -103,13 +103,22 @@ $msbuildArtifactsDir = "$repoFolder/artifacts/msbuild"
 $msbuildLogFilePath = "$msbuildArtifactsDir/msbuild.log"
 $msBuildResponseFile = "$msbuildArtifactsDir/msbuild.rsp"
 
+$preflightClpOption='/clp:DisableConsoleColor'
+$msbuildClpOption='/clp:DisableConsoleColor;Summary'
+if ("${env:CI}${env:APPVEYOR}${env:TEAMCITY_VERSION}${env:TRAVIS}" -eq "")
+{
+    # Not on any of the CI machines. Fine to use colors.
+    $preflightClpOption=''
+    $msbuildClpOption='/clp:Summary'
+}
+
 $msBuildArguments = @"
 /nologo
 /m
 /p:RepositoryRoot="$repoFolder/"
 /fl
 /flp:LogFile="$msbuildLogFilePath";Verbosity=detailed;Encoding=UTF-8
-/clp:Summary
+$msbuildClpOption
 "$makeFileProj"
 "@
 
@@ -122,5 +131,5 @@ if (!(Test-Path $msbuildArtifactsDir))
 
 $msBuildArguments | Out-File -Encoding ASCII -FilePath $msBuildResponseFile
 
-exec dotnet msbuild /nologo /t:Restore /p:PreflightRestore=true /p:RepositoryRoot="$repoFolder/" "$makeFileProj"
+exec dotnet msbuild /nologo $preflightClpOption /t:Restore /p:PreflightRestore=true "$makeFileProj"
 exec dotnet msbuild `@"$msBuildResponseFile"

--- a/build/KoreBuild.sh
+++ b/build/KoreBuild.sh
@@ -16,14 +16,7 @@ __exec() {
 
     local cmdname=$(basename $cmd)
     echo -e "${CYAN}> $cmdname $@${RESET}"
-
-    if [ -z "${TRAVIS}" ]; then
-        $cmd "$@"
-    else
-        # Work around https://github.com/Microsoft/msbuild/issues/1792
-        $cmd "$@" | tee /dev/null
-    fi
-
+    $cmd "$@"
     local exitCode=$?
     if [ $exitCode -ne 0 ]; then
         echo -e "${RED}'$cmdname $@' failed with exit code $exitCode${RESET}" 1>&2
@@ -133,29 +126,14 @@ else
     fi
 fi
 
-# workaround for CLI issue: https://github.com/dotnet/cli/issues/2143
-DOTNET_PATH=`which dotnet | head -n 1`
-ROOT_PATH=`dirname $DOTNET_PATH`
-FOUND=`find $ROOT_PATH/shared -name dotnet`
-if [ ! -z "$FOUND" ]; then
-    echo $FOUND | xargs rm
-fi
-
 netfxversion='4.6.1'
 if [ "$NUGET_PACKAGES" == "" ]; then
     NUGET_PACKAGES="$HOME/.nuget/packages"
 fi
 export ReferenceAssemblyRoot=$NUGET_PACKAGES/netframeworkreferenceassemblies/$netfxversion/content
 
-nugetPath="$scriptRoot/nuget.exe"
-if [ ! -f $nugetPath ]; then
-    nugetUrl="https://dist.nuget.org/win-x86-commandline/v4.0.0-rc4/NuGet.exe"
-    wget -O $nugetPath $nugetUrl 2>/dev/null || curl -o $nugetPath --location $nugetUrl 2>/dev/null
-fi
-
 makeFileProj="$scriptRoot/KoreBuild.proj"
 msbuildArtifactsDir="$repoFolder/artifacts/msbuild"
-msbuildPreflightResponseFile="$msbuildArtifactsDir/msbuild.preflight.rsp"
 msbuildResponseFile="$msbuildArtifactsDir/msbuild.rsp"
 msbuildLogFile="$msbuildArtifactsDir/msbuild.log"
 
@@ -163,38 +141,16 @@ if [ ! -f $msbuildArtifactsDir ]; then
     mkdir -p $msbuildArtifactsDir
 fi
 
-preflightClpOption='/clp:DisableConsoleColor'
-msbuildClpOption='/clp:DisableConsoleColor;Summary'
-if [ -z "${CI}${APPVEYOR}${TEAMCITY_VERSION}${TRAVIS}" ]; then
-    # Not on any of the CI machines. Fine to use colors.
-    preflightClpOption=''
-    msbuildClpOption='/clp:Summary'
-fi
-
-cat > $msbuildPreflightResponseFile <<ENDMSBUILDPREFLIGHT
-/nologo
-/p:NetFxVersion=$netfxversion
-/p:PreflightRestore=true
-/p:RepositoryRoot="$repoFolder/"
-/t:Restore
-$preflightClpOption
-"$makeFileProj"
-ENDMSBUILDPREFLIGHT
-
-# workaround https://github.com/dotnet/core-setup/issues/1664
-echo "{\"sdk\":{\"version\":\"$KOREBUILD_DOTNET_VERSION\"}}" > "$repoFolder/global.json"
-
-__exec dotnet msbuild @"$msbuildPreflightResponseFile"
-
 cat > $msbuildResponseFile <<ENDMSBUILDARGS
 /nologo
 /m
 /p:RepositoryRoot="$repoFolder/"
 /fl
 /flp:LogFile="$msbuildLogFile";Verbosity=detailed;Encoding=UTF-8
-$msbuildClpOption
+/clp:Summary
 "$makeFileProj"
 ENDMSBUILDARGS
 echo -e "$msbuild_args" >> $msbuildResponseFile
 
+__exec dotnet msbuild /nologo /t:Restore /p:PreflightRestore=true /p:NetFxVersion=$netfxversion "$makeFileProj"
 __exec dotnet msbuild @"$msbuildResponseFile"

--- a/build/KoreBuild.sh
+++ b/build/KoreBuild.sh
@@ -16,7 +16,14 @@ __exec() {
 
     local cmdname=$(basename $cmd)
     echo -e "${CYAN}> $cmdname $@${RESET}"
-    $cmd "$@"
+
+    if [ -z "${TRAVIS}" ]; then
+        $cmd "$@"
+    else
+        # Work around https://github.com/Microsoft/msbuild/issues/1792
+        $cmd "$@" | tee /dev/null
+    fi
+
     local exitCode=$?
     if [ $exitCode -ne 0 ]; then
         echo -e "${RED}'$cmdname $@' failed with exit code $exitCode${RESET}" 1>&2
@@ -134,6 +141,7 @@ export ReferenceAssemblyRoot=$NUGET_PACKAGES/netframeworkreferenceassemblies/$ne
 
 makeFileProj="$scriptRoot/KoreBuild.proj"
 msbuildArtifactsDir="$repoFolder/artifacts/msbuild"
+msbuildPreflightResponseFile="$msbuildArtifactsDir/msbuild.preflight.rsp"
 msbuildResponseFile="$msbuildArtifactsDir/msbuild.rsp"
 msbuildLogFile="$msbuildArtifactsDir/msbuild.log"
 
@@ -141,16 +149,35 @@ if [ ! -f $msbuildArtifactsDir ]; then
     mkdir -p $msbuildArtifactsDir
 fi
 
+preflightClpOption='/clp:DisableConsoleColor'
+msbuildClpOption='/clp:DisableConsoleColor;Summary'
+if [ -z "${CI}${APPVEYOR}${TEAMCITY_VERSION}${TRAVIS}" ]; then
+    # Not on any of the CI machines. Fine to use colors.
+    preflightClpOption=''
+    msbuildClpOption='/clp:Summary'
+fi
+
+cat > $msbuildPreflightResponseFile <<ENDMSBUILDPREFLIGHT
+/nologo
+/p:NetFxVersion=$netfxversion
+/p:PreflightRestore=true
+/p:RepositoryRoot="$repoFolder/"
+/t:Restore
+$preflightClpOption
+"$makeFileProj"
+ENDMSBUILDPREFLIGHT
+
+__exec dotnet msbuild @"$msbuildPreflightResponseFile"
+
 cat > $msbuildResponseFile <<ENDMSBUILDARGS
 /nologo
 /m
 /p:RepositoryRoot="$repoFolder/"
 /fl
 /flp:LogFile="$msbuildLogFile";Verbosity=detailed;Encoding=UTF-8
-/clp:Summary
+$msbuildClpOption
 "$makeFileProj"
 ENDMSBUILDARGS
 echo -e "$msbuild_args" >> $msbuildResponseFile
 
-__exec dotnet msbuild /nologo /t:Restore /p:PreflightRestore=true /p:NetFxVersion=$netfxversion /p:RepositoryRoot="$repoFolder/" "$makeFileProj"
 __exec dotnet msbuild @"$msbuildResponseFile"

--- a/build/KoreBuild.sh
+++ b/build/KoreBuild.sh
@@ -152,5 +152,5 @@ cat > $msbuildResponseFile <<ENDMSBUILDARGS
 ENDMSBUILDARGS
 echo -e "$msbuild_args" >> $msbuildResponseFile
 
-__exec dotnet msbuild /nologo /t:Restore /p:PreflightRestore=true /p:NetFxVersion=$netfxversion "$makeFileProj"
+__exec dotnet msbuild /nologo /t:Restore /p:PreflightRestore=true /p:NetFxVersion=$netfxversion /p:RepositoryRoot="$repoFolder/" "$makeFileProj"
 __exec dotnet msbuild @"$msbuildResponseFile"

--- a/template/version.props
+++ b/template/version.props
@@ -1,6 +1,7 @@
 <!-- This file may be overwritten by automation. Only values allowed here are VersionPrefix and VersionSuffix.  -->
 <Project>
     <PropertyGroup>
+        <!-- These values are just an example and not used in build. -->
         <VersionPrefix>1.2.0</VersionPrefix>
         <VersionSuffix>preview1</VersionSuffix>
     </PropertyGroup>

